nucleus/scene.py

```diff
@@ -0,0 +1,147 @@
+    z: float
+    w: float
+
+
+@dataclass
+class CameraParams:
+    position: Point3D
+    heading: Quaternion
+    fx: float
+    fy: float
+    cx: float
+    cy: float
+
+
+@dataclass
+class SceneDatasetItem:
+    url: str
+    type: DatasetItemType
+    reference_id: Optional[str] = None
+    metadata: Optional[dict] = None
+    camera_params: Optional[CameraParams] = None
+
+    @classmethod
+    def from_json(cls, payload: dict):
+        return cls(
+            url=payload.get(URL_KEY, ""),
+            type=payload.get(TYPE_KEY, ""),
+            reference_id=payload.get(REFERENCE_ID_KEY, None),
+            metadata=payload.get(METADATA_KEY, None),
+            camera_params=payload.get(CAMERA_PARAMS_KEY, None),
+        )
+
+    def to_payload(self) -> dict:
+        return {
+            URL_KEY: self.url,
+            TYPE_KEY: self.type,
+            REFERENCE_ID_KEY: self.reference_id,
+            METADATA_KEY: self.metadata,
+            CAMERA_PARAMS_KEY: self.camera_params,
+        }
+
+    def to_json(self) -> str:
+        return json.dumps(self.to_payload(), allow_nan=False)
+
+
+@dataclass
+class Frame:
+    items: Dict[str, SceneDatasetItem] = {}
+
+    def __post_init__(self):
+        for key, value in self.items.items():
+            assert isinstance(key, str), "All keys must be names of sensors"
+            assert isinstance(
+                value, SceneDatasetItem
+            ), "All values must be SceneDatasetItems"
+
+    def add_item(self, item: SceneDatasetItem, sensor_name: str):
+        self.items[sensor_name] = item
+
+    def to_payload(self) -> dict:
+        return {
+            sensor: scene_dataset_item.to_payload()
+            for sensor, scene_dataset_item in self.items.items()
+        }
+
+
+@dataclass
+class Scene:
+    reference_id: str
+    frames: List[Frame] = []
+    metadata: Optional[dict] = None
+
+    # TODO: move validation to scene upload
+    def __post_init__(self):
+        assert isinstance(self.frames, List), "frames must be a list"
+        for frame in self.frames:
+            assert isinstance(
+                frame, Frame
+            ), "each element of frames must be a Frame object"
+        assert len(self.frames) > 0, "frames must have length of at least 1"
+        assert isinstance(
+            self.reference_id, str
+        ), "reference_id must be a string"
+
+    def add_frame(self, frame: Frame):
+        self.frames.append(frame)
+
+    def to_payload(self) -> dict:
+        return {
+            REFERENCE_ID_KEY: self.reference_id,
+            FRAMES_KEY: [frame.to_payload() for frame in self.frames],
+            METADATA_KEY: self.metadata,
+        }
+
+
+@dataclass
+class LidarScene(Scene):
+    def validate(self):
+        lidar_sources = flatten(
+            [
+                [
+                    source
+                    for source in frame.items.keys()
+                    if frame.items[source].type == DatasetItemType.POINTCLOUD
```

Comment:
this is not interpreted as a string comparison for some reason
```suggestion
                   item for frame.items.values() if str(item.type) == DatasetItemType.POINTCLOUD
```

Reply:
actually this check will fail - tested it in my terminal. Since the right hand side is of type `DatasetItemType`, I have to compare `frame.items[source].type` rather than `str()` of that

Reply:
both sides have to be of type `DatasetItemType`, so I casted the left side to `DatasetItemType`

Reply:
changed `type` param to string and perform validation to check one of enum values - resolves issue